crates/ruff_cli/src/diagnostics.rs

```diff
@@ -83,7 +91,25 @@ pub fn lint_path(
     };
 
     // Read the file from disk.
-    let contents = std::fs::read_to_string(path)?;
+    let (contents, jupyter_index) = if is_jupyter_notebook(&path) {
```

Comment:
Idly wondering if we should tag files by dialect in the `resolver`, and pass around tagged files, rather than checking extensions on `Path` everywhere. Same goes for the `.pyi` files. (Maybe not actionable, just a rough thought.)

Reply:
maybe a  `enum FileInfo { Py, PyI, Jupyter { index: JupyterIndex }}`, we could also put that in a hash map `Path` (or string) -> `FileInfo` that we read when collecting the results